crates/frui_widgets/src/basic.rs

```diff
@@ -1,6 +1,64 @@
+use std::ops::Deref;
+
 use frui::prelude::*;
 use frui::render::*;
 
+use crate::{Alignment, BoxLayoutData, Directional, EdgeInsets, TextDirection};
+
+pub trait ChildParentDataProvider<T: RenderWidget> {
+    fn ensure_parent_data<F, P>(&self, ctx: &LayoutCtx<T>, default: F)
+    where
+        F: Fn() -> P,
+        P: 'static;
+}
+
+impl<T: RenderWidget> ChildParentDataProvider<T> for T {
+    fn ensure_parent_data<F, P>(&self, ctx: &LayoutCtx<T>, default: F)
+    where
+        F: Fn() -> P,
+        P: 'static,
+    {
+        for child in ctx.children() {
+            if child.try_parent_data::<P>().is_none() {
+                let data = default();
+                child.set_parent_data(data);
+            }
+        }
+    }
+}
+
+#[derive(InheritedWidget, Builder)]
+pub struct Directionality<T: Widget> {
+    pub direction: TextDirection,
+    pub child: T,
+}
+
+impl<T: Widget> WidgetState for Directionality<T> {
+    type State = TextDirection;
+
+    fn create_state(&self) -> Self::State {
+        self.direction
+    }
+}
+
+impl<T: Widget> InheritedWidget for Directionality<T> {
+    fn build<'w>(&'w self) -> Self::Widget<'w> {
+        &self.child
+    }
+}
+
+impl Directionality<()> {
+    pub fn of<T>(ctx: &LayoutCtx<T>) -> Option<TextDirection> {
+        let state = ctx.depend_on_inherited_widget::<Self>();
+        state.map(|s| *s.as_ref().deref())
+    }
+
+    pub fn of_or_default<T>(ctx: &LayoutCtx<T>) -> TextDirection {
+        Self::of(ctx).unwrap_or_default()
+    }
+    
+}
+
 #[derive(RenderWidget, Builder)]
 pub struct ColoredBox<T: Widget> {
     pub child: T,
```

Comment:
Could you move `ColoredBox` to `frui_widgets/src/boxes.rs`? I think it will make more sense.

Reply:
sure

Reply:
Done